testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/test/helper/hlc/N4CliHelper.java

```diff
@@ -279,36 +293,101 @@ public static void appendExternalOutputToStdout(File outputLogFile) {
+
+		// create yarn workspace
+		if (createYarnWorkspace) {
+			// create package.json
+			List<String> packageJsonLines = Lists.newArrayList(
+					"{",
+					"\t\"private\": true,",
+					"\t\"workspaces\": [ \"packages/*\" ]",
+					"}");
+			Files.write(destinationLocation.resolve(N4JSGlobals.PACKAGE_JSON), packageJsonLines);
+
+			// create node_modules folder
+			Path nodeModulesFolder = destinationLocation.resolve(N4JSGlobals.NODE_MODULES);
+			Files.createDirectories(nodeModulesFolder);
+			for (Path project : Files.list(projectLocation).collect(Collectors.toList())) {
+				if (Files.isDirectory(project)) {
+					Files.createSymbolicLink(
+							nodeModulesFolder.resolve(project.getFileName()),
+							project.toAbsolutePath());
+				}
+			}
+		}
+	}
+
+	/**
+	 * Same as {@link #copyN4jsLibsToLocation(Path, Predicate)}, but the n4js libraries that are to be installed can be
+	 * provided by name instead of a predicate.
+	 */
+	public static void copyN4jsLibsToLocation(Path location, String... n4jsLibs) throws IOException {
+		copyN4jsLibsToLocation(location, libName -> org.eclipse.xtext.util.Arrays.contains(n4jsLibs, libName));
+	}
+
 	/**
 	 * Copies the n4js libraries to the given testing workspace {@code location}.
 	 *
-	 * Only includes n4js libraries (cf. shipped code), for whose project name {@code n4jsLibrariesPredicate} returns
-	 * {@code true}.
+	 * Only includes n4js libraries, for whose project name {@code n4jsLibrariesPredicate} returns {@code true}.
 	 *
 	 * @throws IOException
 	 *             In case the copying is not successful.
 	 */
-	public static void copyN4jsLibsToLocation(File location,
-			Predicate<String> n4jsLibrariesPredicate) throws IOException {
-		// obtain paths of all shipped code projects
-		final List<File> n4jsLibraries = StreamSupport
-				.stream(ShippedCodeAccess.getAllShippedPaths().spliterator(), false)
-				.flatMap(path -> Arrays.asList(new File(path).listFiles()).stream()).collect(Collectors.toList());
-
-		// copy N4JS libraries on demand
-		if (!n4jsLibraries.isEmpty()) {
-			for (final File n4jsLibrary : n4jsLibraries) {
-				if (n4jsLibrariesPredicate.apply(n4jsLibrary.getName())) {
-					if (N4JS_LIBS_BLACKLIST.contains(n4jsLibrary.getName())) {
-						continue;
-					}
-					System.out.println("Including N4JS library in workspace: '" + n4jsLibrary.getName() + "'.");
-					final File libFolder = new File(location, n4jsLibrary.getName());
-					libFolder.mkdir();
-					checkState(libFolder.isDirectory(),
-							"Error while copying N4JS library '" + n4jsLibrary.getName() + "' to workspace.");
-					FileCopier.copy(n4jsLibrary.toPath(), libFolder.toPath(), true);
-				}
+	public static void copyN4jsLibsToLocation(Path location, Predicate<String> n4jsLibrariesPredicate)
+			throws IOException {
+
+		GlobalStateMemento globalState = null;
+		if (!JSONStandaloneSetup.isSetUp()) {
+			globalState = GlobalRegistries.makeCopyOfGlobalState();
+			JSONStandaloneSetup.doSetup();
```

Comment:
Good point. I added this comment in the code:
```
// Restore the original global state (if we had to change it)
// This is important for these cases in particular:
// 1) tests that invoke N4jscBase#doMain() should run *without* any global setup, because #doMain()
// should work if invoked from a plain Java main() method; if we provided the JSON setup as a
// side-effect of this utility method we would "help" the implementation of #doMain() and might overlook
// problems in #doMain()'s own setup.
// 2) some tests deliberately run without any or with an incomplete setup in order to test some failure
// behavior in this broken case.
```